core/src/main/java/hudson/cli/CLIAction.java

```diff
@@ -204,10 +204,17 @@ protected void opened() {
             }
 
             @Override
-            protected void binary(byte[] payload, int offset, int len) {
+            protected void binary(ByteBuffer data) {
                 try {
-                    connection.handle(new DataInputStream(new ByteArrayInputStream(payload, offset, len)));
-                    receivedBytes += len;
+                    if (data.hasArray()) {
+                        connection.handle(new DataInputStream(new ByteArrayInputStream(data.array(), data.arrayOffset() + data.position(), data.remaining())));
+                        receivedBytes += data.remaining();
+                    } else {
+                        byte[] payload = new byte[data.remaining()];
+                        data.get(payload);
+                        connection.handle(new DataInputStream(new ByteArrayInputStream(payload)));
+                        receivedBytes += payload.length;
+                    }
```

Comment:
The CLI requires a `DataInputStream`, so we must convert the `ByteBuffer` into the traditional trio of `byte[] payload`, `int offset`, and `int length`. If the `ByteBuffer` is backed by an array, we can do this without an array copy operation.